
# HG changeset patch
# User alitvinov
# Date 1587407141 -3600
# Node ID ead0b50a6a79ac0803ab226de89462b56afeb80f
# Parent  d5d7e0df08d2b92fbcfbe353b33e43561807eea7
8242498: Invalid "sun.awt.TimedWindowEvent" object leads to JVM crash
Reviewed-by: prr, serb

diff -r d5d7e0df08d2 -r ead0b50a6a79 src/windows/native/sun/windows/awt_Window.cpp
--- a/src/windows/native/sun/windows/awt_Window.cpp	Mon Apr 20 09:35:04 2020 -0700
+++ b/src/windows/native/sun/windows/awt_Window.cpp	Mon Apr 20 19:25:41 2020 +0100
@@ -1571,6 +1571,15 @@
         CHECK_NULL(sequencedEventConst);
     }
 
+    static jclass windowCls = NULL;
+    if (windowCls == NULL) {
+        jclass windowClsLocal = env->FindClass("java/awt/Window");
+        CHECK_NULL(windowClsLocal);
+        windowCls = (jclass)env->NewGlobalRef(windowClsLocal);
+        env->DeleteLocalRef(windowClsLocal);
+        CHECK_NULL(windowCls);
+    }
+
     if (env->EnsureLocalCapacity(3) < 0) {
         return;
     }
@@ -1581,6 +1590,28 @@
         AwtComponent *awtOpposite = AwtComponent::GetComponent(opposite);
         if (awtOpposite != NULL) {
             jOpposite = awtOpposite->GetTarget(env);
+            if ((jOpposite != NULL) &&
+                !env->IsInstanceOf(jOpposite, windowCls)) {
+                env->DeleteLocalRef(jOpposite);
+                jOpposite = NULL;
+
+                HWND parent = AwtComponent::GetTopLevelParentForWindow(opposite);
+                if ((parent != NULL) && (parent != opposite)) {
+                    if (parent == GetHWnd()) {
+                        jOpposite = env->NewLocalRef(target);
+                    } else {
+                        AwtComponent* awtParent = AwtComponent::GetComponent(parent);
+                        if (awtParent != NULL) {
+                            jOpposite = awtParent->GetTarget(env);
+                            if ((jOpposite != NULL) &&
+                                !env->IsInstanceOf(jOpposite, windowCls)) {
+                                env->DeleteLocalRef(jOpposite);
+                                jOpposite = NULL;
+                            }
+                        }
+                    }
+                }
+            }
         }
     }
     jobject event = env->NewObject(wClassEvent, wEventInitMID, target, id,

